File: day_1/One.hs
module One where

finalFloor :: String -> Int
finalFloor "" = 0
finalFloor xs = foldr upOrDown 0 xs

upOrDown :: Char -> Int -> Int
upOrDown ')' floor = floor - 1
upOrDown '(' floor = floor + 1
upOrDown _ _ = error "Not a paren."

basementChar :: String -> Int
basementChar "" = 0
basementChar xs = incrIfNotBasement xs 0 0 where
    incrIfNotBasement [] _ _ = 0
    incrIfNotBasement (paren:parens) currFloor acc
        | move < 0 = acc + 1
        | move >= 0 = incrIfNotBasement parens move (acc + 1)
        where
            move = upOrDown paren currFloor

File: day_1/lua/one.lua
local one = {}

local function up_or_down(char, floor)
    if char == "(" then
        return floor + 1
    elseif char == ")" then
        return floor - 1
    else
        error("Not a paren: " .. char)
    end
end

function one.final_floor(chars)
    local floor = 0

    for i = 1, chars:len() do
        local char = chars:sub(i, i)
        floor = up_or_down(char, floor)
    end

    return floor
end

function one.basement_char(chars)
    local floor = 0

    for i = 1, chars:len() do
        local char = chars:sub(i, i)
        floor = up_or_down(char, floor)

        if floor < 0 then
            return i
        end
    end
end

return one
